scvi/autotune/_autotune.py

```diff
@@ -0,0 +1,346 @@
+import logging
+from typing import Optional, Union
+
+import anndata
+import ray
+import scanpy as sc
+import torch
+from ray import tune
+from ray.tune import CLIReporter, ExperimentAnalysis
+from ray.tune.schedulers import ASHAScheduler
+from ray.tune.schedulers.trial_scheduler import TrialScheduler
+
+from scvi.model.base import BaseModelClass
+
+from ._callbacks import ModelSave, _TuneReportMetricFunctionsCallback
+
+logger = logging.getLogger(__name__)
+
+
+class Autotune:
+    """
+    Hyperparameter tuning using Ray Tune.
+    Parameters
+    ----------
+    adata
+        AnnData object we will tune the model on.
+    model
+        Model from scvi.model we will tune.
+    num_epochs
+        Number of epochs to tune the model over
+    training_metrics
+        Metrics to track during training.
+    metric_functions
+        For metrics calculated after training a model, like silhouette distance.
+    model_hyperparams
+        Config for the model hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    trainer_hyperparams
+        Config for the trainer hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    plan_hyperparams
+        Config for the training_plan hyperparameters https://docs.ray.io/en/master/tune/api_docs/search_space.html.
+    test_effect_covariates
+        Whether to test the effect of including covariates in the model. Default is `True`
+    continious_covariates
+        Continious covariates from `adata` to tune the model on
+    categorical_covariates
+        Categorical covariates from `adata` to tune the model on
+    test_effect_hvg
+        Whether to test the effect of filtering adata with highly variable genes on model performance or not. Default is `False`.
+    top_hvg
+        Top hvgs to test. If `test_effect_hvg` is set, defaults to top 2000, 2500, 3000 and 3500 highly variable genes
+    batch_key
+        Column in adata.obs specifying the different batches in the data.
+
+     Examples
+    --------
+    >>> adata = anndata.read_h5ad(path_to_anndata)
+    >>> tuner = Autotune(adata, scvi.model.SCVI, num_epochs=5)
+    >>> best_model, analysis = tuner.run(metric="elbo_validation")
+    """
+
+    def __init__(
+        self,
+        adata: anndata.AnnData,
+        model: BaseModelClass,
+        num_epochs: int = 2,
+        training_metrics: Optional[list[str]] = None,
+        metric_functions: Optional[dict] = None,
+        model_hyperparams: Optional[dict] = None,
+        trainer_hyperparams: Optional[dict] = None,
+        plan_hyperparams: Optional[dict] = None,
+        continious_covariates: Optional[list[str]] = None,
```

Comment:
These covariates are not present in every model's `setup_anndata`, so we would need to restrict these values to be non-None only in the case that they are present in the signature of the `setup_anndata` method. Additionally, I'm wondering if for kwargs like these covariates or hvg selection below, we should raise warnings if users attempt to autotune over a metric like `elbo_validation`. Since the model architectures would be entirely different, it would not be a fair comparison. However, I can see using the silhouette for batch being a reasonable one.


also there is a typo in: continuous*

Reply:
Good catch. I've included a `self.setup_signature` to the Autotune object that we can use to set covariates to non-None values. I've also added a user warning for the latter (very valid) point you make. Maybe we can consider raising an error if we just don't want to support this behaviour?